### PR TITLE
HTTP semantic convention stability migration for django

### PR DESCRIPTION
**Pull request by lzchen**
Part of https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2453

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -37,15 +37,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2673](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2673))
 - `opentelemetry-instrumentation-django` Add `http.target` to Django duration metric attributes
   ([#2624](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2624))
+  - `opentelemetry-instrumentation-django` Implement new semantic convention opt-in with stable http semantic conventions
```
Comment (emdneto) [inline CHANGELOG.md:40]:
```suggestion
- `opentelemetry-instrumentation-django` Implement new semantic convention opt-in with stable http semantic conventions
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -37,15 +37,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  ([#2714](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2714))
 
 ### Breaking changes
 
 - `opentelemetry-instrumentation-asgi`, `opentelemetry-instrumentation-fastapi`, `opentelemetry-instrumentation-starlette` Use `tracer` and `meter` of originating components instead of one from `asgi` middleware
   ([#2580](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2580))
 - Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `asgi` middleware
   ([#2610](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2610))
-  - Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `fastapi` middleware
+  - Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `fastapi` instrumentation
```
Comment (emdneto) [inline CHANGELOG.md:49]:
```suggestion
- Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `fastapi` instrumentation
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -37,15 +37,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
   ([#2682](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2682))
+    - Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `django` middleware
```
Comment (emdneto) [inline CHANGELOG.md:51]:
```suggestion
- Populate `{method}` as `HTTP` on `_OTHER` methods from scope for `django` middleware
```

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py [hunk 4]
```diff
@@ -517,6 +720,104 @@ def test_wsgi_metrics(self):
...
+        ]
+        _recommended_attrs = {
+            "http.server.active_requests": _server_active_requests_count_attrs_new,
+            "http.server.request.duration": _server_duration_attrs_new,
+        }
+        start = default_timer()
+        for _ in range(3):
+            response = Client().get("/span_name/1234/")
+            self.assertEqual(response.status_code, 200)
+        duration = default_timer() - start
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:737]: can we use `duration_s` like we did in other instrumentation (ex., fastapi) ? just to not confuse anyone

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py [hunk 5]
```diff
@@ -416,13 +421,46 @@ def process_response(self, request, response):
...
-            self._duration_histogram.record(duration, duration_attrs)
+            duration_s = default_timer() - request_start_time
+            if self._duration_histogram_old:
+                self._duration_histogram_old.record(
+                    max(round(duration_s * 1000), 0), duration_attrs_old
+                )
+            if self._duration_histogram_new:
+                self._duration_histogram_new.record(
+                    max(duration_s, 0), duration_attrs_new
+                )
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py:438]:
```suggestion
            duration_s = default_timer() - request_start_time
            if self._duration_histogram_old:
                duration_attrs_old = _parse_duration_attrs(
                    duration_attrs, _HTTPStabilityMode.DEFAULT
                )
                self._duration_histogram_old.record(
                    max(round(duration_s * 1000), 0), duration_attrs_old
                )
            if self._duration_histogram_new:
                duration_attrs_new = _parse_duration_attrs(
                    duration_attrs, _HTTPStabilityMode.HTTP
                )
                self._duration_histogram_new.record(
                    max(duration_s, 0), duration_attrs_new
                )
```

Comment (emdneto) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py:438]: what do you think?

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py [hunk 6]
```diff
@@ -279,6 +409,65 @@ def test_error(self):
             event.attributes[SpanAttributes.EXCEPTION_MESSAGE], "error"
         )
 
+    def test_error_new_semconv(self):
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:412]: Are we missing `self.assertEqual(span.attributes[ERROR_TYPE], '500')` ?

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py [hunk 7]
```diff
@@ -343,6 +532,16 @@ def test_span_name_404(self):
...
 
+    def test_nonstandard_http_method_span_name(self):
+        Client().request(
+            REQUEST_METHOD="NONSTANDARD", PATH_INFO="/span_name/1234/"
+        )
+        span_list = self.memory_exporter.get_finished_spans()
+        self.assertEqual(len(span_list), 1)
+
+        span = span_list[0]
+        self.assertEqual(span.name, "HTTP")
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:543]: I think here would be nice to also assert the http method is `_OTHER`.
Also, as suggestion would be nice to add a test case to check in new semconv where nonstandard http also requires the `http.request.method_original`

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py [hunk 8]
```diff
@@ -517,6 +715,105 @@ def test_wsgi_metrics(self):
...
+                        if isinstance(point, HistogramDataPoint):
+                            self.assertEqual(point.count, 3)
+                            histrogram_data_point_seen = True
+                            if metric.name == "http.server.request.duration":
+                                self.assertAlmostEqual(
+                                    duration_s, point.sum, places=1
+                                )
+                            elif metric.name == "http.server.duration":
+                                self.assertAlmostEqual(
+                                    duration, point.sum, delta=10
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:806]: Any reason why deltas are different from the `new_semconv` and `default_semconv` tests?

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py (hunk 5 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py:438]: Fine with me.

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py [hunk 9]
```diff
@@ -517,6 +749,105 @@ def test_wsgi_metrics(self):
...
+            "http.server.active_requests": active_count_both_attrs,
+            "http.server.request.duration": _server_duration_attrs_new,
+            "http.server.duration": _server_duration_attrs_old,
+        }
+        start = default_timer()
+        for _ in range(3):
+            response = Client().get("/span_name/1234/")
+            self.assertEqual(response.status_code, 200)
+        duration = max(round((default_timer() - start) * 1000), 0)
+        duration_s = max(default_timer() - start, 0)
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py:816]:
```suggestion
        duration_s = max(default_timer() - start, 0)
        duration = max(round(duration_s * 1000), 0)
```

**Review by emdneto — APPROVED** [review]

**Pull request merged**